Add support for WeMo Humidifier

sqldiablo opened the pull request:
## Description:
Adds support for the Holmes WeMo Smart Humidifier device.


**Related issue (if applicable):** None

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4192

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

sqldiablo committed:
WIP - Add WeMo Humidifier

WIP - Changes to add a device to support the WeMo Humidifier device.

-- Bumped version of pywemo
-- Updated base wemo component (wemo.py in root of components folder)
-- Updated base fan device (__init__.py) to support additional speed settings as well as target and current humidity
-- Created wemo.py file for the fan device type (still WIP)

sqldiablo committed:
WIP - Add WeMo Humidifier

WIP - Changes to add a device to support the WeMo Humidifier device.

-- Removed unnecessary high and low target humidity features from fan/__init__.py

sqldiablo committed:
WIP - Add WeMo Humidifier

WIP - Changes to add a device to support the WeMo Humidifier device.

-- Added new supported features to the base fan object for filter life and filter expired
-- Continued work on the wemo humidifier object in the fan component class

sqldiablo committed:
WeMo Humidifier

Various bug fixes to the new WeMo Humidifier device

sqldiablo committed:
WeMo Humidifier

More bugfixes

homeassistant commented:
Hi @sqldiablo, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#11076).

Once you do that we will be able to review and accept this pull request.

Thanks!

sqldiablo commented:
I'm having some trouble getting tox setup, since this is my first time contributing to Home Assistant. I'll work on that and fixing this issues that houndci-bot found this week.

sqldiablo committed:
WeMo Humidifier - Syntax fixes

Fixed syntax issues found by houndci-bot

sqldiablo committed:
WeMo Humidifier - More syntax fixes

sqldiablo committed:
WeMo Humidifier - More syntax fixes

sqldiablo committed:
WeMo Humidifier - More syntax fixes

sqldiablo committed:
WeMo Humidifier - More syntax fixes

sqldiablo committed:
WeMo Humidifier - More syntax fixes

sqldiablo commented:
I was finally able to setup a good dev/test environment and get tox to work. All tests passed on python 3.4. @andrey-git, can you please review this PR and let me know what you think?

sqldiablo commented:
@pavoni & @balloob, would you mind reviewing this PR as well? I'm new to developing for HA, so please let me know if I need to make any changes to better fit with existing standard/precedents.

sqldiablo committed:
WeMo Humidifier - Bump pywemo version

Bump pywemo version due to fix in pywemo humidifier class for better compatibility with Python 2.7

MartinHjelmare commented on homeassistant/components/fan/wemo.py:
```diff
@@ -0,0 +1,248 @@
+        elif WEMO_FAN_MAXIMUM == current_wemo_speed:
+            return SPEED_MAXIMUM
+        else:
+            return STATE_UNKNOWN
```
Return `None` if state is not known.

MartinHjelmare commented on homeassistant/components/fan/wemo.py:
```diff
@@ -0,0 +1,248 @@
+    def speed(self) -> str:
+        """Return the current speed."""
+        current_wemo_speed = self.wemo.fan_mode
+        if WEMO_FAN_OFF == current_wemo_speed:
```
Instead of all the `if ... elif`, make a dict that maps the wemo state to the home assistant state. Define the dict at the module level above. Then lookup that dict here using the wemo speed state, and return the found home assistant state.

MartinHjelmare commented on homeassistant/components/fan/wemo.py:
```diff
@@ -0,0 +1,248 @@
+    @property
+    def speed_list(self: ToggleEntity) -> list:
+        """Get the list of available speeds."""
+        supported_speeds = []
```
Since this is a static list, please define it as a constant at the module level above. Then just return that list here. Please don't append to the list to populate it, just define the list right away.

```py
SUPPORTED_SPEEDS = [
    SPEED_OFF, SPEED_MINIMUM, ...]
```

MartinHjelmare commented on homeassistant/components/fan/wemo.py:
```diff
@@ -0,0 +1,248 @@
+    @property
+    def supported_features(self: ToggleEntity) -> int:
+        """Flag supported features."""
+        return SUPPORT_SET_SPEED | SUPPORT_TARGET_HUMIDITY |\
```
Same as above.

MartinHjelmare reviewed (changes requested):
Hi! Sorry about the long delay for a review.

Since this PR adds a service and more state attributes to the base fan component, we need at least another opinion from a member, if this change is ok, or if the service and new state attributes should start in the wemo fan platform only. Having new features starting in the platform is the usual way when new services etc are added that are not obviously a feature for all platforms of the component.

See below for my other comments.

sqldiablo commented on homeassistant/components/fan/wemo.py:
```diff
@@ -0,0 +1,248 @@
+WEMO_WATER_GOOD = 2
+
+
+# pylint: disable=unused-argument, too-many-function-args
```
I copied much of this code from existing platforms (used the dyson fan and other wemo components as models for this platform, so these were things that I "inherited" from them. I'll clean this up.

sqldiablo committed:
WeMo Humidifier - Fixes from code review

Various fixes to the WeMo Humidifier platform, as requested for PR 11076

sqldiablo committed:
WeMo Humidifier - More fixes from code review

Various fixes to the WeMo Humidifier platform, as requested for PR 11076

sqldiablo committed:
WeMo Humidifier - More fixes from code review

Various fixes to the WeMo Humidifier platform, as requested for PR 11076

sqldiablo committed:
WeMo Humidifier - More fixes from code review

Various fixes to the WeMo Humidifier platform, as requested for PR 11076

sqldiablo commented on homeassistant/components/fan/wemo.py:
```diff
@@ -0,0 +1,222 @@
+        # look up model name once as it incurs network traffic
+        self._model_name = self.wemo.model_name
+
+    @asyncio.coroutine
```
I'm not sure why this is still showing up, but this works just fine in my local home assistant install.

MartinHjelmare replied:
U think you can ignore this. Hound is a bit slow sometimes.

sqldiablo commented:
I've made the changes requested, and functional testing in my local install works great.

sqldiablo committed:
WeMo Humidifier - More fixes from code review

Fixed whitespace issues

MartinHjelmare commented on homeassistant/components/fan/wemo.py:
```diff
@@ -0,0 +1,224 @@
+https://home-assistant.io/components/fan.wemo/
+"""
+import logging
+import asyncio
```
Please sort the imports 🔡, within each group. Ie put asyncio before logging.

MartinHjelmare commented on homeassistant/components/fan/wemo.py:
```diff
@@ -0,0 +1,223 @@
+
+    def set_speed(self: ToggleEntity, speed: str) -> None:
+        """Set the fan_mode of the Humidifier."""
+        if speed == SPEED_OFF:
```
You could use a dict here too, which is the inverse of the other state dict, that maps the home assistant speed to the wemo speed.

MartinHjelmare commented on homeassistant/components/fan/wemo.py:
```diff
@@ -0,0 +1,223 @@
+    @property
+    def speed(self) -> str:
+        """Return the current speed."""
+        return WEMO_FAN_SPEED_TO_HASS.get(self.wemo.fan_mode, None)
```
`None` is the default value returned if the key is missing.

MartinHjelmare reviewed (commented):
Looks good in general. Some minor comments below.

We still need a consensus on adding the new service etc in the fan base component.

sqldiablo committed:
WeMo Humidifier - More fixes from code review

Added another dict for lookups and re-ordered imports.

sqldiablo committed:
WeMo Humidifier - More fixes from code review

Corrected indentation issues.

MartinHjelmare commented:
If you don't want to wait for a review from another member regarding the new service etc, in the fan base component, my suggestion is to move all of that into the wemo fan platform. Look eg at the sonos media_player platform how a platform can register services.

If more fan platforms come along that require the humidity service we can move it to the base fan component at that time.

The above is the only thing in my view that's holding this off from being merged.

sqldiablo commented:
That's not a bad idea. I don't suspect many fans have/need humidity control. I'm out of town right now, so I'll work on that next week.

-- Adam

(Sent from my mobile device)
________________________________
From: Martin Hjelmare <notifications@github.com>
Sent: Saturday, December 30, 2017 6:39 PM
To: home-assistant/home-assistant
Cc: Adam C. Belebczuk; Mention
Subject: Re: [home-assistant/home-assistant] Add support for WeMo Humidifier (#11076)


If you don't want to wait for a review from another member regarding the new service etc, in the fan base component, my suggestion is to move all of that into the wemo fan platform. Look eg at the sonos media_player platform how a platform can register services.

If more fan platforms come along that require the humidity service we can move it to the base fan component at that time.

The above is the only thing in my view that's holding this off from being merged.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/home-assistant/pull/11076#issuecomment-354575059>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABn-cm2rR1YQXF7wvNtGKzemq-RjBDGaks5tFsmpgaJpZM4Q84O->.

sqldiablo committed:
Reset fork to allow merge from upstream

sqldiablo committed:
Merge remote-tracking branch 'upstream/dev' into dev

balloob commented:
I think you broke something in your branch, 0 files changed?

sqldiablo commented:
The changes in 61.1 broke my changes in this PR, so I pulled out all of my changes and stashed them somewhere local. I then merged upstream dev into my fork so I can have a clean start. Once I get some time, I'll work on the changes that Martin requested (not adding anything to __init__.py and keeping it all in my wemo.py fan implementation) and also fixing compatibility issues with 61.1.

sqldiablo commented:
So yeah, for now this is PR with no changes.

sqldiablo committed:
Merge pull request #1 from home-assistant/dev

Merge main dev to fork's dev

balloob commented:
This PR seems to have gone stale. Closing it. You can reopen it when you're ready to finish it.
